tests/trx_generator/trx_provider.hpp

```diff
@@ -18,45 +19,88 @@ namespace eosio::testing {
 
-      logged_trx_data(eosio::chain::transaction_id_type trx_id, fc::time_point sent=fc::time_point::now()) :
+      explicit logged_trx_data(eosio::chain::transaction_id_type trx_id, fc::time_point sent=fc::time_point::now()) :
          _trx_id(trx_id), _sent_timestamp(sent) {}
    };
 
    struct provider_base_config {
-      std::string _peer_endpoint = "127.0.0.1";
-      unsigned short _port = 9876;
+      std::string    _peer_endpoint_type = "p2p";
+      std::string    _peer_endpoint      = "127.0.0.1";
+      unsigned short _port               = 9876;
 
       std::string to_string() const {
          std::ostringstream ss;
-         ss << "peer_endpoint: " << _peer_endpoint << " port: " << _port;
+         ss << "endpoint type: " << _peer_endpoint_type << "peer_endpoint: " << _peer_endpoint << " port: " << _port;
          return std::move(ss).str();
       }
    };
 
-   struct p2p_connection {
-      const provider_base_config& _config;
-      boost::asio::io_service _p2p_service;
+   struct provider_connection {
+      const provider_base_config&                                 _config;
+      eosio::chain::named_thread_pool<struct provider_connection> _connection_thread_pool;
+
+      explicit provider_connection(const provider_base_config& provider_config)
+          : _config(provider_config) {}
+
+      virtual ~provider_connection() = default;
+
+      void init_and_connect() {
+         _connection_thread_pool.start(
+             1, [](const fc::exception& e) { elog("provider_connection exception ${e}", ("e", e)); });
+         connect();
+      };
+
+      void cleanup_and_disconnect() {
+         disconnect();
+         _connection_thread_pool.stop();
+      };
+
+      virtual void send_transaction(const chain::packed_transaction& trx) = 0;
+
+    private:
+      virtual void connect()    = 0;
+      virtual void disconnect() = 0;
+   };
+
+   struct http_connection : public provider_connection {
+      std::atomic<uint64_t> _acknowledged{0};
+      std::atomic<uint64_t> _sent{0};
+
+      explicit http_connection(const provider_base_config& provider_config)
+          : provider_connection(provider_config) {}
+
+      void send_transaction(const chain::packed_transaction& trx);
+
+    private:
+      void connect() final;
+      void disconnect() final;
+   };
+
+   struct p2p_connection : public provider_connection {
       boost::asio::ip::tcp::socket _p2p_socket;
 
-      p2p_connection(const provider_base_config& provider_config) :
-            _config(provider_config), _p2p_service(), _p2p_socket(_p2p_service) {}
+      explicit p2p_connection(const provider_base_config& provider_config)
+          : provider_connection(provider_config)
+          , _p2p_socket(_connection_thread_pool.get_executor()) {}
 
-      void connect();
-      void disconnect();
       void send_transaction(const chain::packed_transaction& trx);
+
+    private:
+      void connect() final;
+      void disconnect() final;
```

Comment:
I like `override`, it helps if someone changes the parent class.
```suggestion
      void connect() override final;
      void disconnect() override final;
```

Reply:
Addressed: [28e8d55](https://github.com/AntelopeIO/leap/pull/1138/commits/28e8d5518991526b66ce68411d1bd309fd3498d7)